### PR TITLE
Feature/18 professor survey UI

### PR DESCRIPTION
**Pull request by Ahnaf-Ahmed**
Changes
---
Create a survey view for a professor to fill in their course preferences.
Includes a subset of data in classList.json to simulate the /courses endpoint to retrieve classes

Testing
---
Manual testing

Linked Issue
---
https://github.com/SENG499-Company-4/Frontend/issues/18

Additional Info
---
Sends data to console for now. Can be changed to end point when integrating
image
image

### CODE REVIEW

File: app/src/classList.json [hunk 1]
```diff
@@ -0,0 +1,975 @@
+[
```
Comment (bmazerolle) [inline app/src/classList.json:1]: You should be using the existing data instead to get your classes (from my previous PR) since that's what you'll be getting from API anyways. I assume you pulled all of this from that file, so you should just be able to add that into the `Survey.tsx` file as a hook that just extracts this bit of the total returned data object. You'll feed in the same info and we won't have to add the conversion later

File: app/src/components/organisms/SurveyClassQuestion.tsx [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      return {
+        ...willingRadio,
+        [event.target.value]: true
+      };
+    });
+
+    props.fieldChanged(props.name, 'willing', (event.target as HTMLInputElement).value);
+  };
+
+  return (
```
Comment (bmazerolle) [inline app/src/components/organisms/SurveyClassQuestion.tsx:47]: I think we are missing the element of those courses where the professor _has_ to have a specific qualification like a PENG. I think the "can't teach" radio implies this, but we should be more specific imo. I think the backend _should_ store this information but I dont think they are right now so we'll need to get it. For now we could add an asterisk to certain courses and show in the UI that an asterisk means additional qualifications are necessary.

File: app/src/components/organisms/SurveyClassQuestion.tsx [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    <Stack key={props.name} direction="row" spacing={5}>
```
Comment (bmazerolle) [inline app/src/components/organisms/SurveyClassQuestion.tsx:48]: UI is a good start but I think we would benefit by cutting down on the vertical space for each row. I would add a new column to the left that contains the course number (ie. CSC 423). Then, add headers to the grids for Course number | Teaching ability | Willingness to teach

File: app/src/components/organisms/SurveyForm.tsx [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        currentValues.explanation = event.target.value;
+      }
+      return currentValues;
+    });
+  };
+
+  return (
+    <div>
+      <form onSubmit={onSubmit}>
+        <Stack style={{ height: '500px', overflowY: 'scroll' }} spacing={4}>
```
Comment (bmazerolle) [inline app/src/components/organisms/SurveyForm.tsx:59]: I would avoid using a static height, when I minimize my browser this overflows. You could either add a bottom margin on this stack with enough space for a static height bottom stack (the relief stuff), or just set it to have height 75% to keep it simple.

File: app/src/components/organisms/SurveyForm.tsx [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+          {Object.keys(values.courses).map((field) => {
+            return <SurveyClassQuestion key={field} name={field} fieldChanged={fieldChanged} />;
+          })}
+        </Stack>
+        <Divider />
+        <Stack spacing={4}>
+          <FormControlLabel
+            control={<Checkbox />}
+            label="Have Relief?"
```
Comment (bmazerolle) [inline app/src/components/organisms/SurveyForm.tsx:68]: I don't understand what this question means, probably needs more detail. I see that you also disable the corresponding text entry unless this is checked which is great, but I would also hide the box `display:none` unless the box is checked.

File: app/src/components/organisms/SurveyForm.tsx [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+          <TextField
+            id="Explanation-textarea"
+            label="Explanation"
+            multiline
+            rows={4}
+            disabled={disable}
+            style={{ width: 500 }}
+            inputProps={{ style: { color: 'black' } }}
+            onChange={(event) => handleCheckbox(event as React.ChangeEvent<HTMLInputElement>, false)}
+          />
```
Comment (bmazerolle) [inline app/src/components/organisms/SurveyForm.tsx:80]: I think we're also missing form entries for "Which semesters are you available (spring/summer/fall)" and "How many courses do you want to teach"

File: app/src/components/organisms/SurveyForm.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            label="Have Relief?"
+            onChange={(event) => handleCheckbox(event as React.ChangeEvent<HTMLInputElement>, true)}
+          />
+          <TextField
+            id="Explanation-textarea"
+            label="Explanation"
+            multiline
+            rows={4}
+            disabled={disable}
+            style={{ width: 500 }}
```
Comment (bmazerolle) [inline app/src/components/organisms/SurveyForm.tsx:77]: Same as above comment, avoid fixed width/height stuff. Maybe make this a % width

File: app/src/components/organisms/SurveyClassQuestion.tsx [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+  name: string;
+  fieldChanged: (fieldId: string, type: string, value: string) => void;
+}) {
+  const [disabled, setDisabled] = useState(true);
+  const [checked, setChecked] = useState({ willing: false, notWilling: false, veryWilling: false });
+
+  const handleDisabling = (event: React.ChangeEvent<HTMLInputElement>) => {
+    if (event.target.value === ability.cannot) {
+      setDisabled(event.target.checked);
+
```
Comment (bmazerolle) [inline app/src/components/organisms/SurveyClassQuestion.tsx:20]: Unnecessary whitespace

**Review by bmazerolle — CHANGES_REQUESTED** [review]
Nice job @Ahnaf-Ahmed, some UI tweaks but this looks good. Good choice on the shared/ dir with the interfaces too.

File: app/src/components/organisms/SurveyForm.tsx (hunk 6 above)
Comment (bmazerolle) [reply app/src/components/organisms/SurveyForm.tsx:80]: Also "What days do you want to teach" and "how many classes max per day"

File: app/src/components/organisms/SurveyClassQuestion.tsx [hunk 9]
```diff
@@ -0,0 +1,106 @@
+import React, { useState } from 'react';
+import Radio from '@mui/material/Radio';
+import RadioGroup from '@mui/material/RadioGroup';
+import FormControlLabel from '@mui/material/FormControlLabel';
+import FormControl from '@mui/material/FormControl';
+import FormLabel from '@mui/material/FormLabel';
+import Stack from '@mui/material/Stack';
+import { willingRadio, ability, willing } from '../shared/constants/surveyForm.constants';
```
Comment (sam-warren) [inline app/src/components/organisms/SurveyClassQuestion.tsx:8]: Let's change this relative import to an absolute one ('components/shared/constants/surveyForm.constants';

File: app/src/components/organisms/SurveyClassQuestion.tsx [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+  fieldChanged: (fieldId: string, type: string, value: string) => void;
+}) {
+  const [disabled, setDisabled] = useState(true);
+  const [checked, setChecked] = useState({ willing: false, notWilling: false, veryWilling: false });
+
+  const handleDisabling = (event: React.ChangeEvent<HTMLInputElement>) => {
+    if (event.target.value === ability.cannot) {
+      setDisabled(event.target.checked);
+
+      setChecked(() => {
```
Comment (sam-warren) [inline app/src/components/organisms/SurveyClassQuestion.tsx:21]: Can this be condensed to `setChecked(willingRadio)`?

File: app/src/components/organisms/SurveyClassQuestion.tsx (hunk 3 above)
Comment (sam-warren) [reply app/src/components/organisms/SurveyClassQuestion.tsx:48]: I agree - spacing={2} might be a cleaner look

File: app/src/components/organisms/SurveyForm.tsx (hunk 5 above)
Comment (sam-warren) [reply app/src/components/organisms/SurveyForm.tsx:68]: We could also wrap this field in a <Tooltip> to provide extra information without cluttering the page.

File: app/src/components/pages/Survey.tsx [hunk 11]
```diff
@@ -1,7 +1,18 @@
 import React from 'react';
+import SurveyForm from 'components/organisms/SurveyForm';
+import Typography from '@mui/material/Typography';
+import myData from '../../classList.json';
```
Comment (sam-warren) [inline app/src/components/pages/Survey.tsx:4]: Relative import should be absolute, can just import using `import myData from 'classList.json'`

File: app/src/components/shared/constants/surveyForm.constants.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
+export const willingRadio = {
```
Comment (sam-warren) [inline app/src/components/shared/constants/surveyForm.constants.ts:1]: I like what you're doing here - would these be better suited as `enum`s than constants?

File: app/src/components/shared/interfaces/surveyForm.interfaces.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
+export interface course {
```
Comment (sam-warren) [inline app/src/components/shared/interfaces/surveyForm.interfaces.ts:1]: I also have some interfaces built up for this stuff but I suspect your PR will be approved and merged first so let's chat once Hexuan's PR is ready and we'll sort out our interfaces. Nice work!

File: app/src/classList.json (hunk 1 above)
Comment (Ahnaf-Ahmed) [reply app/src/classList.json:1]: Since the existing data is outside the src folder I don't think I can access it. I made that file and trimmed it down to signify what it was for, then I was thinking we would delete it when we integrate with api

File: app/src/components/organisms/SurveyClassQuestion.tsx (hunk 2 above)
Comment (Ahnaf-Ahmed) [reply app/src/components/organisms/SurveyClassQuestion.tsx:47]: Added a additionalQualifications prop to surveyClassQuestion that just adds required to the radio. This doesn't actually do anything other than display an asterisk

File: app/src/components/pages/Survey.tsx [hunk 14]
```diff
@@ -1,7 +1,18 @@
 import React from 'react';
+import SurveyForm from 'components/organisms/SurveyForm';
+import Typography from '@mui/material/Typography';
+import myData from '../../classList.json';
 
 function Survey() {
-  return <div>Survey</div>;
+  return (
+    <div>
```
[the remark this answers is not in the document]
Comment (Ahnaf-Ahmed) [reply app/src/components/pages/Survey.tsx:8]: Changed it to flexbox and center aligned

File: app/src/components/organisms/SurveyClassQuestion.tsx (hunk 3 above)
Comment (Ahnaf-Ahmed) [reply app/src/components/organisms/SurveyClassQuestion.tsx:48]: The spacing for each row is in the surveyForm component and is now decreased to 2 from 4

File: app/src/components/organisms/SurveyForm.tsx (hunk 4 above)
Comment (Ahnaf-Ahmed) [reply app/src/components/organisms/SurveyForm.tsx:59]: height is now using a calculated value that should be 50%. Overflow does not like it when a percentage height is used so calc() was needed.

File: app/src/components/organisms/SurveyForm.tsx (hunk 5 above)
Comment (Ahnaf-Ahmed) [reply app/src/components/organisms/SurveyForm.tsx:68]: Had a conversation with Ben about this. Course Relief is essentially a pass to take one less course per relief. Changed to accepting a number rather than a checkbox to reflect this.

File: app/src/components/organisms/SurveyForm.tsx (hunk 6 above)
Comment (Ahnaf-Ahmed) [reply app/src/components/organisms/SurveyForm.tsx:80]: Has now been added. How many courses do you want to teach isnt in requirements and is instead determnined by whether they are research or teaching profs. Dropdown for this has been added

**Review by sam-warren — APPROVED** [review]

**Review by bmazerolle — APPROVED** [review]
Reviewed again and tested it out, nice job
